### PR TITLE
arrayplex setitem & refactoring

### PR DESCRIPTION
**Pull request by tigercosmos**
- moved shared code  to `array_common.hpp`
- supported arrayplex to set items.

### CODE REVIEW

File: cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArray.cpp]: use the function from `array_common.hpp`, same for other functions.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp]: use the function from `array_common.hpp`

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp:289]: support set items for arrayplex.

File: tests/test_buffer.py
Comment (tigercosmos) [inline tests/test_buffer.py:840]: it works :)

**Comment by tigercosmos** [comment]
@yungyuc  the PR is ready for review.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/array_common.hpp]: `shape_type` should not be in `modmesh::python` namespace as it has been in `modmesh::detail` namespace.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/array_common.hpp]: I don't like `slice_type` to be a global alias, but I am not sure how to deal with it yet.  You can leave it here.  I will think more about it.

Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: It's good to factor out the standalone common file.  It makes issues like this stand out.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/array_common.hpp]: The multitude of free function (templates) made me think that maybe classes for shape and slice would further simplify the implementation.  What do you think?

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp]: This (and alike) are good.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp]: This is too magical.  The original lambda is much more readable.  Please don't use this overly tricky template code.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp]: Lambda is clearer

File: tests/test_buffer.py
Comment (yungyuc) [reply tests/test_buffer.py:840]: Fantastic

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] Limit the scope of `shape_type`
- [x] Discuss: whether or not to use classes for shape and slice
- [x] Avoid overly tricky template code

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: I agree with you. A helper class might be useful.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp]: I see. I thought it is readable if both wrapper class call the same function.
So, do you want to revert the change for both "shape" and "stride"?

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: maybe we can also put it `modmesh::detail`. I couldn't find a good place for it. Maybe `SimpleArray.hpp` is fine, but it might be better to have a header file for basic types.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: However, if the existing functions lack a robust correlation, the helper class might just serve as a namespace.

**Comment by yungyuc** [comment]
@tigercosmos any update?

**Comment by tigercosmos** [comment]
> @tigercosmos any update?

@yungyuc  I've left few comments few days ago, and waiting for your response.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: I am looking for a good design pattern for this these days, so far I haven't found it yet.

Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: Do you want to use this PR to make them classes, or leave it here for a future PR?  I am fine either way.  We should choose one way to proceed quickly.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp]: I think the original lambda is good for this occurrence.  Specifically here reverting is good.

In other occurrences maybe more review is needed.  I am fine to use this PR or another one to handle the cases that need review.  A decision should be made quickly.

**Comment by yungyuc** [comment]
Thanks, @tigercosmos .  I added comments inline.

Please always leave a global comment after you have done your commenting and wait for feedback, like what I am doing in this comment.

Inline comments are hard to track.  For example, in the following snapshot, the only clue for me to tell whether I should review again is the message you left 32 minutes ago:

image

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: let's make it in another PR.

File: cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp]: reverted.

**Comment by tigercosmos** [comment]
@yungyuc  regarding the issues above:
 - Limit the scope of shape_type: Done
- Discuss: whether or not to use classes for shape and slice:  Done
- Avoid overly tricky template code: Reverted.

**Comment by tigercosmos** [comment]
CI failed due to the temporary failure of macOS-13

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/array_common.hpp:58]: new introduced.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/array_common.hpp:60]: only public APIs are here.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [inline cpp/modmesh/buffer/pymod/array_common.hpp:161]: private functions within the class

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp:58]: no need to repeat template now.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: `shape_type` is still in `modmesh::python` namespace.  This should be fixed.  Please remove this alias When using `shape_type`, spelling out the full scope `modmesh::detail::shape_type`.

Perhaps `slice_type` should be treated in the same way.

File: cpp/modmesh/buffer/SimpleArray.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/SimpleArray.hpp]: While moved it here, it becomes obvious that `int` does not look right.  `slice_type` should be interchangeable with `sshape_type`.  Please review and make them consistent.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Thanks, @tigercosmos .  There are an remaining issue and a new one:

- [x] (Still need to address) Limit the scope of `shape_type`
- [x] Review `slice_type` and its consistency with other type aliases

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/array_common.hpp]: Why do you need `inline` for functions already in the class declaration?

Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: I want to make the function code to be inline within the caller. I think static and inline can be used at the same time?

File: cpp/modmesh/buffer/SimpleArray.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/SimpleArray.hpp]: thanks for catching this. fixed.

**Comment by tigercosmos** [comment]
@yungyuc fixed the issue of `slice_type` and `shape_type`, and added comments.

File: cpp/modmesh/buffer/SimpleArray.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/SimpleArray.hpp]: Why do you use an unsized integer rather than a signed?  Slice can be negative-valued.

File: cpp/modmesh/buffer/pymod/TypeBroadcast.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/TypeBroadcast.hpp:43]: It's OK to have type aliases in the class.  Why removing them?

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: They can, but the `inline` should be no op, because code in the class declaration is automatically inlined, IIRC.

**Review by yungyuc — CHANGES_REQUESTED** [review]
One more issue in your update:

- [ ] Signness of the value_type of slice_type

File: cpp/modmesh/buffer/SimpleArray.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/SimpleArray.hpp:87]: I see, then I think your previous comment, https://github.com/solvcon/modmesh/pull/303#discussion_r1557555171, should not be an issue?

shape is always positive, and slice can be negative, so they don't need to be interchangeable.

File: cpp/modmesh/buffer/pymod/TypeBroadcast.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/TypeBroadcast.hpp:43]: I misunderstood. I thought you were asking to remove all aliases.

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/array_common.hpp]: sure, removed.

**Comment by tigercosmos** [comment]
@yungyuc please check the comments.

File: cpp/modmesh/buffer/SimpleArray.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/SimpleArray.hpp]: I misunderstood, fixed.

File: cpp/modmesh/buffer/pymod/TypeBroadcast.hpp
Comment (tigercosmos) [reply cpp/modmesh/buffer/pymod/TypeBroadcast.hpp:43]: reverted.

**Comment by tigercosmos** [comment]
CI failed due to unrelated issues for all macOS

**Comment by tigercosmos** [comment]
@yungyuc  fixed all issues, ready for review.

File: cpp/modmesh/buffer/SimpleArray.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/SimpleArray.hpp:87]: Great

File: cpp/modmesh/buffer/pymod/array_common.hpp
Comment (yungyuc) [inline cpp/modmesh/buffer/pymod/array_common.hpp:64]: Looks good

File: cpp/modmesh/buffer/pymod/TypeBroadcast.hpp
Comment (yungyuc) [reply cpp/modmesh/buffer/pymod/TypeBroadcast.hpp:43]: No problem

**Review by yungyuc — APPROVED** [review]
LGTM.  The CI mac runs are unstable.  I will merge once the CI are back online.

**Pull request merged**
